src/networking.c

```diff
@@ -3188,11 +3196,17 @@ int checkClientOutputBufferLimits(client *c) {
      * like normal clients. */
     if (class == CLIENT_TYPE_MASTER) class = CLIENT_TYPE_NORMAL;
 
-    if (server.client_obuf_limits[class].hard_limit_bytes &&
-        used_mem >= server.client_obuf_limits[class].hard_limit_bytes)
+    /* convert limits from negative values to % of maxmemory if needed */
+    long long hard_limit_bytes = server.client_obuf_limits[class].hard_limit >= 0 ?
+        server.client_obuf_limits[class].hard_limit:
+        server.client_obuf_limits[class].hard_limit * (long long)server.maxmemory / -100;
+    long long soft_limit_bytes = server.client_obuf_limits[class].soft_limit >= 0 ?
+        server.client_obuf_limits[class].soft_limit:
+        server.client_obuf_limits[class].soft_limit * (long long)server.maxmemory / -100;
```

Comment:
I know that it doesn't make a difference, but (to me) it would be more readable to do this:
```
-server.client_obuf_limits[class].soft_limit * (long long)server.maxmemory / 100;
```
Putting the `-` in from of the `soft_limit` value would seem to emphasize that this is the negative value we want to change the sign on.

FWIW, on these ternaries, I think they look more readable like this:
```
value = (parenthesized expression)
    ? if-true-value
    : if-false-value
```

Reply:
you mean put the `?` and `:` in the beginning of the line?
it's nice for indentation, but i think conceptually the `?` belongs to the first line?
anyway, i don't see we have any occurrence of this style now in redis, so i also don't wanna mix styles.

Reply:
i refactored this code, it should be much more clearer in the new form, hope it's enough.